src/codeFlowLogin.ts

```diff
@@ -61,7 +62,72 @@ export async function checkRedirectServer(adfs: boolean) {
 
 let terminateServer: () => Promise<void>;
 
+function parseQuery(uri: vscode.Uri) {
+	return uri.query.split('&').reduce((prev: any, current) => {
+		const queryString = current.split('=');
+		prev[queryString[0]] = queryString[1];
+		return prev;
+	}, {});
+}
+
+class UriEventHandler extends vscode.EventEmitter<vscode.Uri> implements vscode.UriHandler {
+	public handleUri(uri: vscode.Uri) {
+		this.fire(uri);
+	}
+}
+
+const handler = new UriEventHandler();
+
+vscode.window.registerUriHandler(handler);
+
+async function exchangeCodeForToken(clientId: string, environment: AzureEnvironment, tenantId: string, callbackUri: string, state: string) {
+	let uriEventListener: vscode.Disposable;
+	return new Promise((resolve: (value: TokenResponse) => void , reject) => {
+		uriEventListener = handler.event(async (uri: vscode.Uri) => {
+			try {
+				const query = parseQuery(uri);
+				const code = query.code;
+	
+				if (query.state !== state) {
+					throw new Error('State does not match.');
+				}
+	
+				resolve(await tokenWithAuthorizationCode(clientId, environment, callbackUri, tenantId, code));
+			} catch (err) {
+				reject(err);
+			}
+		});
+	}).then(result => {
+		uriEventListener.dispose()
+		return result;
+	}).catch(err => {
+		uriEventListener.dispose();
+		throw err;
+	});
+}
+
+async function loginWithoutLocalServer(clientId: string, environment: AzureEnvironment, adfs: boolean, tenantId: string): Promise<TokenResponse> {
+	const callbackUri = (await vscode.env.createAppUri()).toString();
+	const nonce = crypto.randomBytes(16).toString('base64');
+	const signInUrl = `${environment.activeDirectoryEndpointUrl}${adfs ? '' : `${tenantId}/`}oauth2/authorize?response_type=code&client_id=${encodeURIComponent(clientId)}&redirect_uri=${encodeURIComponent(callbackUri)}&state=${nonce}&resource=${encodeURIComponent(environment.activeDirectoryResourceId)}&prompt=select_account`;
```

Comment:
When the callbackUri is `http://localhost:port/path` you can use `redirectUrlAAD` and encode the port in the state parameter like below. That server will then redirect to localhost. (This helps with the fact that you cannot register a `http` URL as a redirect URL, not even for localhost.)

Reply:
For ADFS just use the callbackUri for now. Each ADFS instance has its own allow-list of redirect URIs and that includes the port number whereas for AAD the hostname is enough. So ADFS might not work in the web for now. I think we should wait for someone to have the need for this to work first before investing time.